src/levanter/optim/adamh.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2025 The Levanter Authors
+# SPDX-License-Identifier: Apache-2.0
+
+import dataclasses
+from dataclasses import dataclass
+from typing import NamedTuple, Optional, Any
+
+import chex
+import jax
+import jax.numpy as jnp
+import optax
+from optax import tree_utils as otu
+
+from haliax.nn import Linear
+
+from levanter.optim.config import OptimizerConfig
+from levanter.utils.jax_utils import leaf_key_paths
+
+
+@OptimizerConfig.register_subclass("adamH")
+@dataclass(frozen=True)
+class AdamHConfig(OptimizerConfig):
+    """
+    This is a variant of the Adam optimizer configuration.
+
+    We ensure that the linear weights stay exactly constant norm as initialization by applying the following update rule:
+
+    p_new_intermediate = p - learning_rate * u * norm(p) / norm(u)
+    p_new = p_new_intermediate / norm(p_new_intermediate) * norm(p)
```

Comment:
big fan of this idea! (I still want to try (softly) moment matching the initial gaussian)